### PR TITLE
[#202] adding new txn scaledown unit tests

### LINKED ISSUE #202: Enhance transaction recovery scaled down unit test coverage
The transaction scale down processing is quite complex and the operator code misses a bit a better coverage in tests.
Adding at least one full scale down unit test would be nice.

On top of that other tests on other scenarios could be evolved and the new code could be required to provide a test scenario.

### PR DESCRIPTION
**Pull request by ochaloup**
fixes #202 

@yersan I'm adding some more unit tests of the txn scale down processing.

### CODE REVIEW

**Review by yersan — CHANGES_REQUESTED** [review]
Hello @ochaloup , I've found minor cosmetic changes and one issue with one of the implemented tests. Could you take a look? Thanks

File: pkg/controller/wildflyserver/transaction_recovery_test.go [hunk 1]
```diff
@@ -123,19 +160,159 @@ func TestWildFlyServerControllerScaleDown(t *testing.T) {
...
+	// Reconcile for the scale down - updating the pod labels
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+
+	// Mocking the jboss-cli.sh calls to return of there is no transactions subsystem available
+	wildflyutil.RemoteOps = &remoteOpsMock{executeReturnString: []string{
+		`{"outcome": "success", "result": ["nothing"]}`, // does not contain "transactions"
+	}}
+
+	// Reconcile for the txn scale down procesing - recovery skipped
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply pkg/controller/wildflyserver/transaction_recovery_test.go]: +1

File: pkg/controller/wildflyserver/transaction_recovery_test.go [hunk 2]
```diff
@@ -102,18 +126,31 @@ func TestWildFlyServerControllerScaleDown(t *testing.T) {
...
+	assert.Equal(expectedReplicaSize, wildflyServer.Status.Replicas)
+}
+
+func TestRecoveryScaleDownToPodInvestigation(t *testing.T) {
+	wildflyServer := defaultWildflyServerDefinition.DeepCopy()
+	expectedReplicaSize := int32(1)
+
+	setupBeforeScaleDown(t, wildflyServer, expectedReplicaSize)
+
+	log.Info("WildFly server was reconciliated to the state the pod status corresponds with namespace. Let's scale it down.",
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply pkg/controller/wildflyserver/transaction_recovery_test.go]: +1

File: pkg/controller/wildflyserver/transaction_recovery_test.go [hunk 3]
```diff
@@ -123,19 +160,159 @@ func TestWildFlyServerControllerScaleDown(t *testing.T) {
...
+	wildflyServer.Spec.Replicas = 0
+	err := cl.Update(context.TODO(), wildflyServer)
+
+	// Reconcile for the scale down - updating the pod labels
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+
+	// Mocking the jboss-cli.sh calls to reach the loop on working with pods at "processTransactionRecoveryScaleDown"
+	wildflyutil.RemoteOps = &remoteOpsMock{executeReturnString: []string{
+		`{"outcome": "success", "result": ["transactions"], "action": "list subsystesm"}`,
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply pkg/controller/wildflyserver/transaction_recovery_test.go]: +1

File: pkg/controller/wildflyserver/transaction_recovery_test.go [hunk 4]
```diff
@@ -123,19 +160,159 @@ func TestWildFlyServerControllerScaleDown(t *testing.T) {
...
+	}}
+
+	// Reconcile for the txn scale down procesing - recovery skipped
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+	// StatefulSet needs to be updated in scaled down manner
+	statefulSet := &appsv1.StatefulSet{}
+	err = cl.Get(context.TODO(), req.NamespacedName, statefulSet)
+	assert.Equal(int32(0), *statefulSet.Spec.Replicas)
+	// WildFlyServer status needs to be updated in sclaed down manner
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply pkg/controller/wildflyserver/transaction_recovery_test.go]: +1

File: pkg/controller/wildflyserver/transaction_recovery_test.go [hunk 5]
```diff
@@ -123,19 +160,159 @@ func TestWildFlyServerControllerScaleDown(t *testing.T) {
...
+	log.Info("WildFly server was reconciliated, let's scale it down.", "WildflyServer", wildflyServer)
+	wildflyServer.Spec.Replicas = 0
+	err := cl.Update(context.TODO(), wildflyServer)
+
+	// Reconcile for the scale down - updating the pod labels
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+
+	// Mocking the jboss-cli.sh calls to return that JDBC object store is not used
+	wildflyutil.RemoteOps = &remoteOpsMock{executeReturnString: []string{
```
[the remark this answers is not in the document]
Comment (ochaloup) [reply pkg/controller/wildflyserver/transaction_recovery_test.go]: good point, thank you. I tried to enhance the assertions in this area.

**Comment by ochaloup** [comment]
@yersan thank you for your review. I fixed the code based on your suggestions and I tried to enhance the handling a bit.

**Review by yersan — APPROVED** [review]

**Comment by yersan** [comment]
@ochaloup thank you!

**Pull request merged**